README.md

## Features

- receive Datagrams on demand without subscription.
- receive timeout.
- Various shorthand methods.

## Usage

usage example:

```dart
import 'dart:convert';
import 'package:easy_udp/easy_udp.dart';

start_server() async {
  // Create a EasyUDPSocket and bind to localhost:7777,
  // Note that you can also manually create a RawDatagramSocket 
  // and pass it to EasyUDPSocket(..) to create a EasyUDPSocket.
  final socket = await EasyUDPSocket.bind('localhost', 7777);

  while (true) {
    // Rather than subscribing to a stream of RawSocketEvents and
    // calling receive in the callback, which is the case when using
    // RawDatagramSocket, with EasyUDPSocket, you can get Datagram 
    // on demand with `receive` method.
    final datagram = await socket.receive();
    print('Server received: ${ascii.decode(datagram.data)} from ${datagram.port}');

    // use `sendBack` to send message to where a Datagram comes from.
    // This is a shorthand of:
    // socket.send(somedata, datagram.address, datagram.port);
    socket.sendBack(datagram, ascii.encode('pong'));
  }
}

start_client(int port) async {
  final socket = await EasyUDPSocket.bind('localhost', port);
  socket.send(ascii.encode('ping'), 'localhost', 7777);
  final resp = await socket.receive();
  print('Client $port received: ${ascii.decode(resp.data)}');

  // `close` method of EasyUDPSocket is awaitable.
  await socket.close();
  print('Client $port closed');
}

main() async {
  start_server();
  await Future.delayed(Duration(seconds: 1));
  start_client(8001);
  start_client(8002);
  start_client(8003);
}

// Output:
//
// Server received: ping from 8001
// Server received: ping from 8002
// Server received: ping from 8003
// Client 8001 received: pong
// Client 8002 received: pong
// Client 8003 received: pong
// Client 8001 closed
// Client 8002 closed
// Client 8003 closed

```

with timeout:
```dart
...
final datagram = await socket.receive(timeout: 1000);
if(datagram == null) {
  print('Receive timeout');
  continue;
}
...

// Output:
//
// Server received: ping from 8001
// Server received: ping from 8002
// Server received: ping from 8003
// Client 8001 received: pong
// Client 8002 received: pong
// Client 8003 received: pong
// Client 8001 closed
// Client 8002 closed
// Client 8003 closed
// Receive timeout
// Receive timeout
// Receive timeout
// Receive timeout

```

## Features and bugs

Please file feature requests and bugs at the [issue tracker][tracker].

[tracker]: http://github.com/xtyxtyx/easy_udp/issues


CHANGELOG.md
## 1.2.0

- add explode option

## 1.1.0

- add timeout feature

## 1.0.0

- automatically do ns lookup in `send`


## 0.1.0+1

- Update documentation

## 0.1.0

- Initial version, created by Stagehand


test/easy_udp_test.dart
import 'package:easy_udp/easy_udp.dart';
import 'package:test/test.dart';

void main() {
  group('A group of tests', () {
    // Awesome awesome;

    setUp(() {
      // awesome = Awesome();
    });

    test('First Test', () {
      // expect(awesome.isAwesome, isTrue);
    });
  });
}


lib/easy_udp.dart
/// usage example:
/// 
/// ```dart
/// import 'dart:convert';
/// import 'package:easy_udp/easy_udp.dart';
/// 
/// start_server() async {
///   // Create a EasyUDPSocket and bind to localhost:7777,
///   // Note that you can also manually create a RawDatagramSocket 
///   // and pass it to EasyUDPSocket(..) to create a EasyUDPSocket.
///   final socket = await EasyUDPSocket.bind('localhost', 7777);
///   print(socket.rawSocket.port);
/// 
///   while (true) {
///     // Rather than subscribing to a stream of RawSocketEvents and
///     // calling receive in the callback, which is the case when using
///     // RawDatagramSocket, with EasyUDPSocket, you can get Datagram 
///     // on demand with `receive` method.
///     final datagram = await socket.receive();
///     print('Server received: ${ascii.decode(datagram.data)} from ${datagram.port}');
/// 
///     // use `sendBack` to send message to where a Datagram comes from.
///     // This is a shorthand of:
///     // socket.send(somedata, datagram.address, datagram.port);
///     socket.sendBack(datagram, ascii.encode('pong'));
///   }
/// }
/// 
/// start_client(int port) async {
///   final socket = await EasyUDPSocket.bind('localhost', port);
///   socket.send(ascii.encode('ping'), 'localhost', 7777);
///   final resp = await socket.receive();
///   print('Client $port received: ${ascii.decode(resp.data)}');
/// 
///   // `close` method of EasyUDPSocket is awaitable.
///   await socket.close();
///   print('Client $port closed');
/// }
/// 
/// main() async {
///   start_server();
///   await Future.delayed(Duration(seconds: 1));
///   start_client(8001);
///   start_client(8002);
///   start_client(8003);
/// }
/// 
/// ````
library easy_udp;

export 'src/easy_udp_base.dart';


lib/src/easy_udp_base.dart
import 'dart:async';
import 'dart:io';
import 'package:async/async.dart';

/// EasyUDPSocket is a wrapper over RawDatagramSocket
/// to make life easier.
class EasyUDPSocket {
  final RawDatagramSocket rawSocket;
  final StreamQueue _eventQueue;

  /// create an EasyUDPSocket with a RawDatagramSocket
  EasyUDPSocket(this.rawSocket) : _eventQueue = StreamQueue(rawSocket);

  /// create an EasyUDPSocket and bind to host:port
  static Future<EasyUDPSocket> bind(dynamic host, int port,
      {bool reuseAddress = true, bool reusePort = false, int ttl = 1}) async {
    final socket = await RawDatagramSocket.bind(host, port,
        reuseAddress: reuseAddress, reusePort: reusePort, ttl: ttl);
    return EasyUDPSocket(socket);
  }

  /// create an EasyUDPSocket and bind to random port.
  static Future<EasyUDPSocket> bindRandom(dynamic host,
      {bool reuseAddress = true, bool reusePort = false, int ttl = 1}) {
    return bind(host, 0,
        reuseAddress: reuseAddress, reusePort: reusePort, ttl: ttl);
  }

  /// receive a Datagram from the socket.
  Future<Datagram> receive({int timeout, bool explode = false}) {
    final completer = Completer<Datagram>.sync();
    if (timeout != null) {
      Future.delayed(Duration(milliseconds: timeout)).then((_) {
        if (!completer.isCompleted) {
          if (explode) {
            completer.completeError('EasyUDP: Receive Timeout');
          } else {
            completer.complete(null);
          }
        }
      });
    }
    Future.microtask(() async {
      while ((await _eventQueue.peek) != RawSocketEvent.read) {
        await _eventQueue.next;
      }
      if (!completer.isCompleted) {
        await _eventQueue.next;
        completer.complete(rawSocket.receive());
      }
    });
    return completer.future;
  }

  /// send some data with this socket.
  Future<int> send(List<int> buffer, dynamic address, int port) async {
    InternetAddress addr;
    if (address is InternetAddress) {
      addr = address;
    } else if (address is String) {
      addr = (await InternetAddress.lookup(address))[0];
    } else {
      throw 'address must be either an InternetAddress or a String';
    }
    return rawSocket.send(buffer, addr, port);
  }

  /// use `sendBack` to send message to where a Datagram comes from.
  /// This is a shorthand of socket.send(somedata, datagram.address, datagram.port);
  int sendBack(Datagram datagram, List<int> buffer) {
    return rawSocket.send(buffer, datagram.address, datagram.port);
  }

  /// close the socket.
  Future<void> close() async {
    rawSocket.close();
    while ((await _eventQueue.next) != RawSocketEvent.closed) {
      continue;
    }
  }
}


example/easy_udp_example.dart
import 'dart:convert';
import 'package:easy_udp/easy_udp.dart';

start_server() async {
  // Create a EasyUDPSocket and bind to localhost:7777,
  // Note that you can also manually create a RawDatagramSocket 
  // and pass it to EasyUDPSocket(..) to create a EasyUDPSocket.
  final socket = await EasyUDPSocket.bind('localhost', 7777);

  while (true) {
    // Rather than subscribing to a stream of RawSocketEvents and
    // calling receive in the callback, which is the case when using
    // RawDatagramSocket, with EasyUDPSocket, you can get Datagram 
    // on demand with `receive` method.
    final datagram = await socket.receive(timeout: 1000);
    if(datagram == null) {
      print('Receive timeout');
      continue;
    }
    print('Server received: ${ascii.decode(datagram.data)} from ${datagram.port}');

    // use `sendBack` to send message to where a Datagram comes from.
    // This is a shorthand of:
    // socket.send(somedata, datagram.address, datagram.port);
    socket.sendBack(datagram, ascii.encode('pong'));
  }
}

start_client(int port) async {
  final socket = await EasyUDPSocket.bind('localhost', port);
  await socket.send(ascii.encode('ping'), 'localhost', 7777);
  final resp = await socket.receive();
  print('Client $port received: ${ascii.decode(resp.data)}');

  // `close` method of EasyUDPSocket is awaitable.
  await socket.close();
  print('Client $port closed');
}

main() async {
  start_server();
  await Future.delayed(Duration(seconds: 1));
  start_client(8001);
  start_client(8002);
  start_client(8003);
}
